include/seahorn/Expr/EvalModel.hh

```diff
@@ -0,0 +1,367 @@
+#pragma once
+
+#include "seahorn/Expr/EvalUtils.hh"
+#include "seahorn/Expr/ExprCore.hh"
+#include "seahorn/Expr/ExprGmp.hh"
+#include "seahorn/Expr/ExprOpBind.hh"
+#include "seahorn/Expr/ExprOpBinder.hh"
+
+#include <math.h>
+#include <stdlib.h>
+
+namespace expr {
+namespace eval {
+
+/**
+ * \class NumContainerBase base class to store numbers
+ *
+ * \details Stores a number and provides methods to get/set the width of the
+ * number
+ */
+template <typename T> struct NumContainerBase {
+  T num;
+
+  NumContainerBase(T numArg, unsigned widthArg) : num(numArg) {}
+
+  virtual void setWidth(unsigned widthArg) = 0;
+  virtual unsigned getWidth() const = 0;
+
+  virtual bool operator<(const NumContainerBase<T> &container) const {
+    return num < container.num;
+  }
+
+  virtual bool operator==(const NumContainerBase<T> &container) const {
+    return (num == container.num);
+  }
+};
+
+/**
+ * Stores primitive number types (unsigned int, ...) and mpz_class
+ *
+ * The width is stored as a separate member variable
+ */
+template <typename T> struct PrimitiveContainer : NumContainerBase<T> {
+  unsigned width;
+
+  PrimitiveContainer(T numArg, unsigned widthArg)
+      : NumContainerBase<T>(numArg, widthArg), width(widthArg) {}
+
+  virtual void setWidth(unsigned widthArg) override { width = widthArg; }
+
+  virtual unsigned getWidth() const override { return width; }
+
+  virtual bool operator==(const NumContainerBase<T> &container) const override {
+    return (this->num == container.num) &&
+           (width ==
+            (static_cast<const PrimitiveContainer<T> &>(container)).width);
+  }
+};
+
+/**
+ * Bit vector structure. Acts as an interface for NumContainerBase which holds a
+ * number and provides methods to get/set the numbers width
+ *
+ * To Add a new type of number (a new NumContainer), template specialize
+ * chooseContainer() with the NumContainer
+ *
+ * \note bools are stored as 1/0 with a width of 1
+ */
+template <typename T> struct BvNum {
+private:
+  std::shared_ptr<NumContainerBase<T>> num;
```

Comment:
this is way too complicated and inefficient. I was proposing not keeping `width` in `BvNum<T>`, and then add it in the specialization `BvNum<mpz_class>`.

The current solution is even heavier than the previous one since now every number contains a pointer on top of the width!

Reply:
Ok it should be fixed now. I remove the BvNum<T> struct all together and stored the number directly. I figured if its not storing the width then its no different than just storing the number directly